### ISSUE TITLE
Dangling data tuples in table oc_cards_properties

### ISSUE THREAD

**Issue by raoulbhatia** [nextcloud/server#15286]
I have a long upgrade history, where I started out on OwnCloud and then migrated over to Nextcloud.  I have also used the Contacts app for a long time.

I was today doing some contacts cleanup and noticed dangling data tuples with the following command:
```SQL
SELECT * FROM oc_cards_properties
LEFT JOIN oc_cards ON oc_cards_properties.cardid = oc_cards.id
HAVING oc_cards.id IS NULL;

38 rows in set (0.00 sec)
```

How about adding a constraint to disallow such invalid data?

Thanks,
Raoul

**Comment by kesselb** [comment]
cc @rullzer @nickvergessen

**Comment by nickvergessen** [comment]
We have a repair step for that in place, which should already take care of this:
https://github.com/nextcloud/server/blob/master/apps/dav/lib/Migration/RemoveOrphanEventsAndContacts.php

it's executed on each update of the dav app. So I guess we can close this?

**Comment by raoulbhatia** [comment]
I suppose this will ship with Nextcloud 16 then, right?

**Comment by nickvergessen** [comment]
It is shipping since Nextcloud 13

**Comment by raoulbhatia** [comment]
Strangely, I am unable to locate this file in my installation.  Also, if this is shipping since NC 13, then I should not be in this situation with my NC 15 installation, right?

```
# ls ./apps/dav/lib/Migration/RemoveOrphanEventsAndContacts.php
ls: cannot access './apps/dav/lib/Migration/RemoveOrphanEventsAndContacts.php': No such file or directory
```
```
# grep -i orphan ./apps/dav/lib/Migration/*
#
```
Also, the history shows that this file was added in January 2019, only?  Check
https://github.com/nextcloud/server/commits/master/apps/dav/lib/Migration/RemoveOrphanEventsAndContacts.php and https://github.com/nextcloud/server/commit/f88e7648d51036c4c04f6c44a8286b1762bea506.patch

**Comment by nickvergessen** [comment]
Oh, you are right. I though this was much longer ago.
So yeah it is fixed in 16. Let's make it available in 15 too.

**Comment by MorrisJobke** [comment]
Fixed for 15 by #15387.
